text/0000-const-generics.md

```diff
@@ -0,0 +1,287 @@
+that depend on a variable) we cannot compare the equality of their values
+because their values are determined by a const variable, the value of which is
+unknown prior to monomorphization.
+
+For this reason we will (initially, at least) treat the return value of const
+expressions as *projections* - values determined by the input, but which are
+not themselves known. This is similar to how we treat associated types today.
+When comparing the evaluation of an abstract const expression - which we'll
+call a *const projection* - to another const of the same type, its equality is
+always unknown.
+
+Therefore we can neither unify nor guarantee the nonunification of any const
+projection with any other const unless they are *syntactically identical.* That
+is, because we require that const equality is reflexive, we know that `{N + 1}`
+is equal to `{N + 1}`, but we don't know whether or not it is equal to
+`{N * 2}` or even to `N` or `{1 + N}`.
+
+#### Future extensions
+
+Someday we could introduce knowledge of the basic properties of some operations
+- such as the commutativity of addition and multiplication - to begin making
+smarter judgments on the equality of const projections. However, this RFC does
+not proposing building any knowledge of that sort into the language and doing
+so would require a future RFC.
+
+## Specialization on const parameters
+
+It is also necessary for specialization that const parameters have a defined
+ordering of specificity. For this purpose, literals are defined as more
+specific than other expressions, otherwise expressions have an indeterminate
+ordering.
+
+Just as we could some day support more advanced notions of equality between
+const projections, we could some day support more advanced definitions of
+specificity. For example, given the type `(i32, i32)`, we could determine that
+`(0, PARAM2)` is more specific than `(PARAM1, PARAM2)` - roughly the analog
+of understanding that `(i32, U)` is more specific than the type `(T, U)`. We
+could also someday support intersectional and other more advanced definitions
+of specialization on constants.
+
+# How We Teach This
+[how-we-teach-this]: #how-we-teach-this
+
+Const generics is a large feature, and will require significant educational
+materials - it will need to be documented in both the book and the reference,
+and will probably need its own section in the book. Documenting const generics
+will be a big project in itself.
+
+However, const generics should be treated as an advanced feature, and it should
+not be something we expose to new users early in their use of Rust.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This feature adds a significant amount of complexity to the type system,
+allowing types to be determined by constants. It requires determining the rules
+around abstract const equality, which result in surprising edge cases. It adds
+a lot of syntax to the language. The language would definitely be simpler if we
```

Comment:
I think we need to work on a shorthand for statements like this, because obviously most new features are going to add additional complexity *somewhere*. It's just noise.